qiskit/transpiler/passes/mapping/algorithm/flexlayer_heuristics.py

```diff
@@ -0,0 +1,408 @@
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""
+A core algorithm of the flexible-layer swap mapping heuristics first proposed in [1].
+
+The outline of the algorithm is as follows.
+0. Assume an initial_layout is given and set it to `layout`.
+1. Initialize `blocking_gates` as gates without in-edge in dependency graph.
+2. Update `blocking_gates` by processing applicable gates for a current `layout`.
+3. If it comes to no blocking gates, it terminates. Otherwise, it selects a qubit
+  pair (= an edge in the coupling graph) to be swapped based on its `cost`.
+4. Add the swap gate at the min-cost edge (= update `layout`).
+5. Go back to the step 2.
+Note: In the actual flow, there is an additional path for avoiding handling cyclic swaps in step 3.
+
+[1] T. Itoko, R. Raymond, T. Imamichi, A. Matsuo, and A. W. Cross.
+Quantum circuit compilers using gate commutation rules.
+In Proceedings of ASP-DAC, pp. 191--196. ACM, 2019.
+
+See https://arxiv.org/abs/1907.02686 (extended version of [1]) for the details of the algorithm.
+"""
+import collections
+import copy
+import logging
+import pprint
+from typing import List
+
+from qiskit.circuit import QuantumCircuit, QuantumRegister
+from qiskit.dagcircuit import DAGCircuit
+from qiskit.extensions.standard import SwapGate
+from qiskit.transpiler.coupling import CouplingMap
+from qiskit.transpiler.exceptions import TranspilerError
+from qiskit.transpiler.layout import Layout
+
+from .dependency_graph import DependencyGraph
+
+logger = logging.getLogger(__name__)
+
+
+class FlexlayerHeuristics:
+    """
+    A core algorithm class implementing the flexible-layer swap mapping heuristics.
+    """
+
+    def __init__(self,
+                 qc: QuantumCircuit,
+                 dependency_graph: DependencyGraph,
+                 coupling: CouplingMap,
+                 initial_layout: Layout,
+                 lookahead_depth: int = 10,
+                 decay_rate: float = 0.5):
+        """
+        Initialize a FlexlayerHeuristics instance.
+        Args:
+            qc: A circuit to be mapped.
+            dependency_graph: Dependency graph of the circuit.
+            coupling: CouplingMap of the target backend.
+            initial_layout: Initial layout (layout at the beginning of circuit).
+            lookahead_depth: How far gates from blocking gates should be looked ahead.
+            decay_rate: Decay rate of look-ahead weight (0 < decay_rate < 1).
+        Raises:
+            TranspilerError: if no (or invalid) initial_layout is given.
+        """
+        self._qc = qc
+
+        if initial_layout is None:
```

Comment:
a swapper should work with "physical circuits" and no layout should be interpreted at this point. This is something that was decided after this PR was initially submitted... sorry about that :(